### PR TITLE
(#66813) Adding support for slots in subclasses of Module.

### PR DESCRIPTION
**Pull request by urviljivani007**
Fixes #66813

### CODE REVIEW

**Review by mikaylagawarecki — COMMENTED** [review]
Perhaps we should also update the `__getstate__` and `__setstate__` for our layers that inherit from `torch.nn.Module` but do not call `super().__getstate__` (I think [RNN](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/rnn.py#L291) is the only one)

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1642,10 +1642,21 @@ def _call_impl(self, *args, **kwargs):
 
     def __getstate__(self):
         state = self.__dict__.copy()
+
+        # Support for slots in child classes
+        for key in self.__slots__:
```
Comment (mikaylagawarecki) [inline torch/nn/modules/module.py:1647]: from the CI failures, which can be seen [here](https://hud.pytorch.org/pr/108017) it looks like this should be gated with a `if hasattr(self, __slots__)`.

Seems like by default, classes do not have a `__slots__` unless it is explicitly defined.

e.g. 

```
class A:
    pass

b = A()
# This fails AttributeError: type object 'A' has no attribute '__slots__'. Did you mean: '__class__'?
# print(A.__slots__)
print(A.__dict__)
```

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -1642,10 +1642,21 @@ def _call_impl(self, *args, **kwargs):
...
+            state[key] = getattr(self, key)
+
         state.pop("_compiled_call_impl", None)
         return state
 
     def __setstate__(self, state):
+
+        # Support for slots in child classes
+        for key in self.__slots__:
+            self.__setattr__(key, state[key])
```
[the remark this answers is not in the document]
Comment (urviljivani007) [reply torch/nn/modules/module.py:1657]: done

File: torch/nn/modules/module.py (hunk 1 above)
Comment (urviljivani007) [reply torch/nn/modules/module.py:1647]: done

File: test/package/test_save_load.py [hunk 3]
```diff
@@ -261,6 +261,23 @@ def test_save_imported_module_using_package_importer(self):
...
+        obj = TestModule()
+        obj.var = "test_value"
+
+        with tempfile.TemporaryFile() as f:
+            torch.save(obj, f)
+            f.seek(0)
+            m_obj = torch.load(f)
+            self.assertEqual(m_obj.var, "test_value")
+
+# Had to move it out as serialization fails on local object
```
[the remark this answers is not in the document]
Comment (urviljivani007) [reply test/package/test_save_load.py:278]: yes, your deduction was correct.

File: test/package/test_save_load.py [hunk 4]
```diff
@@ -261,6 +261,23 @@ def test_save_imported_module_using_package_importer(self):
             exporter.intern("**")
             exporter.save_module("package_a.use_torch_package_importer")
 
+    def test_save_load_with_slots_in_child_class(self):
```
[the remark this answers is not in the document]
Comment (urviljivani007) [reply test/package/test_save_load.py:264]: done.

File: torch/nn/modules/module.py (hunk 1 above)
Comment (urviljivani007) [reply torch/nn/modules/module.py]: @mikaylagawarecki I have addressed all the comments can you please review this code again?

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -1642,10 +1642,24 @@ def _call_impl(self, *args, **kwargs):
 
     def __getstate__(self):
         state = self.__dict__.copy()
+
+        # Support for slots in child classes
+        if getattr(self, "__slots__", None):
```
Comment (mikaylagawarecki) [inline torch/nn/modules/module.py:1647]: I had missed something in my previous pass of the PR, we don't only want to save the `__slots__` defined in the current class, but if this class subclasses other classes that subclass `torch.nn.Module`, we want to get their `__slots__`, which we can do using `copyreg._slotnames`. See below for an example illustrating this:

e.g. 

```python
class Block(torch.nn.Module):
    __slots__ = ['b']
    def __init__(self):
        super().__init__()
        self.b = 3

class DummyModel(Block):
    __slots__ = ['a']
    def __init__(self):
        super().__init__()
        self.a = 2


m = DummyModel()
print(m.__slots__)  # ['a']
print(copyreg._slotnames(m.__class__))  # ['a', 'b']

```

This shows how we do something similar for `torch.Tensor`

https://github.com/pytorch/pytorch/blob/fb288aa99b1c7c1792a3f90396d2ab7275eab2d9/torch/_tensor.py#L202-L205

File: test/package/test_save_load.py [hunk 6]
```diff
@@ -261,6 +261,5 @@ def test_save_imported_module_using_package_importer(self):
             exporter.intern("**")
             exporter.save_module("package_a.use_torch_package_importer")
 
-
```
Comment (mikaylagawarecki) [inline test/package/test_save_load.py:264]: nit: oops :)

**Review by mikaylagawarecki — COMMENTED** [review]
Thank you for your patience and iteration on this PR! There was something I missed regarding slots for Module subclasses in the previous pass that I left a comment on

**Pull request closed without merging**